File: application/backend-credit-card-enrollment/backend-java/src/main/java/cloud/ambar/creditcard/enrollment/projection/enrollmentlist/GetEnrollmentList.java
package cloud.ambar.creditcard.enrollment.projection.enrollmentlist;

import cloud.ambar.creditcard.enrollment.aggregate.EnrollmentStatus;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.web.context.annotation.RequestScope;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequestScope
@RequiredArgsConstructor
public class GetEnrollmentList {
    private final EnrollmentRepository enrollmentRepository;
    private final ProductNameRepository productNameRepository;
    public List<EnrollmentListItem> getEnrollmentList(final String userId) {
        return enrollmentRepository.findAllByUserId(userId)
                .stream()
                .map(enrollment -> EnrollmentListItem.builder()
                        .id(enrollment.getId())
                        .userId(enrollment.getUserId())
                        .productId(enrollment.getProductId())
                        .productName(
                                productNameRepository.findOneById(enrollment.getProductId())
                                        .orElseThrow()
                                        .getName())
                        .requestedDate(enrollment.getRequestedDate())
                        .status(enrollment.getStatus())
                        .statusReason(enrollment.getStatusReason())
                        .reviewedOn(enrollment.getReviewedOn())
                        .build())
                .collect(Collectors.toList());
    }

    public boolean isThereAnyAcceptedEnrollmentForUserAndProduct(final String userId, final String productId) {
        return enrollmentRepository.findAllByUserId(userId)
                .stream()
                .anyMatch(enrollment -> enrollment.getProductId().equals(productId) && enrollment.getStatus().equals(EnrollmentStatus.ACCEPTED.name()));
    }
}

File: application/backend-credit-card-enrollment/backend-c#/CreditCard/Enrollment/Projection/EnrollmentList/GetEnrollmentList.cs
using CreditCardEnrollment.CreditCard.Enrollment.Aggregate;

namespace CreditCardEnrollment.CreditCard.Enrollment.Projection.EnrollmentList;

public class GetEnrollmentList {
    private readonly EnrollmentRepository _enrollmentRepository;
    private readonly ProductNameRepository _productNameRepository;

    public GetEnrollmentList(
        EnrollmentRepository enrollmentRepository,
        ProductNameRepository productNameRepository) {
        _enrollmentRepository = enrollmentRepository;
        _productNameRepository = productNameRepository;
    }

    public IEnumerable<EnrollmentListItem> GetList(string userId) {
        return _enrollmentRepository.FindAllByUserId(userId)
            .Select(enrollment => new EnrollmentListItem {
                Id = enrollment.Id,
                UserId = enrollment.UserId,
                ProductId = enrollment.ProductId,
                ProductName = _productNameRepository.FindOneById(enrollment.ProductId)?.Name 
                              ?? throw new InvalidOperationException(),
                RequestedDate = enrollment.RequestedDate,
                Status = enrollment.Status,
                StatusReason = enrollment.StatusReason,
                ReviewedOn = enrollment.ReviewedOn
            });
    }

    public bool IsThereAnyAcceptedEnrollmentForUserAndProduct(string userId, string productId) {
        return _enrollmentRepository.FindAllByUserId(userId)
            .Any(e => e.ProductId == productId && e.Status == EnrollmentStatus.Accepted.ToString());
    }
}